syntax = "proto3";

package wide_eye;

import "google/protobuf/timestamp.proto";

// Service for client â†” dispatcher communication
service ClientDispatcher {
  // User registration
  rpc Register (RegisterRequest) returns (RegisterResponse);

  // User login
  rpc Login    (LoginRequest)    returns (LoginResponse);

  // Start a new task
  rpc StartTask (TaskRequest)    returns (TaskStartResponse);

  // Stream results for a given task
  rpc StreamResults (TaskResultsRequest) returns (stream TaskResult);

  // Send available categories and locations
  rpc ListAvailableCategories(ListCategoriesRequest) returns (ListCategoriesResponse);
  rpc ListAvailableLocations(ListLocationsRequest) returns (ListLocationsResponse);

}

// --- Request / Response Messages ---

// Registration
message RegisterRequest {
  string username = 1;
  string password = 2;
}

message RegisterResponse {
  bool   success = 1;
  string message = 2;   // e.g., "Registration successful" or error text
  string user_id = 3;   // Assigned user identifier
}

// Login
message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  bool   success = 1;
  string message = 2;   // e.g., "Login successful" or error text
  string token   = 3;   // Auth token for session
}

// Task creation
message TaskRequest {
  string                     token      = 1; // Auth token from Login
  string                     keywords   = 2;
  string                     categories = 3;
  string                     location   = 4;
  google.protobuf.Timestamp  start_time = 5;
  google.protobuf.Timestamp  end_time   = 6;
}

message TaskStartResponse {
  bool   success = 1;
  string message = 2;   // e.g., "Task started"
  string task_id = 3;   // Unique ID assigned by dispatcher
}

// Requesting a results stream
message TaskResultsRequest {
  string token   = 1;
  string task_id = 2;
}

// Individual streamed result
message TaskResult {
  string                     task_id   = 1;
  string                     title     = 2;
  string                     url       = 3;
  google.protobuf.Timestamp  timestamp = 4;
}

// Available category/location update
// Empty request (you can use google.protobuf.Empty if you want later)
message ListCategoriesRequest {}

message ListCategoriesResponse {
    repeated string categories = 1;
}

message ListLocationsRequest {}

message ListLocationsResponse {
    repeated string locations = 1;
}
